---
title: "Looking at Gene Distribution"
author: "Craig McLean"
date: "4/5/2020"
output: html_document
---

# Goal

I'd like to determine if we can categorize genes or pathways into separate 
classes.

```{r}
pathways <- read.csv(here::here("data/2020-04-05/Oman_kegg_mapp_lcms.csv"))
x <- as.matrix(pathways[,-c(1)])
rownames(x) <- pathways$Function
x <- x[rowSums(x) > 0,]
```

```{r, coordinating the NMF}
library(NMF)

if(T) {
    res <- nmfEstimateRank(x, seq(2,5), method='brunet', 
                       nrun=100, 
                       seed=123456)
    rV <- randomize(x)
    rand <- nmfEstimateRank(rV[rowSums(rV) > 0,], seq(2,5), 
                            method='brunet', 
                            nrun=100, 
                            seed=123456)
    
    png(filename = here::here("results/2020-04-05/metagenomes_cophenetic_cor.png"),
        width = 6.5, height = 5, res = 300, units = "in")
    plot(res, rand)
    dev.off()
}
```

Looks like k = 3 is the winner for the metagenomes.

```{r}
nmfOut <- nmf(x, rank = 3)

wMatrix <- basis(nmfOut)
    
heatmap(wMatrix)

keyFeatures <- extractFeatures(nmfOut, method = 0.99)
linkedFactors <- apply(coef(nmfOut), 1, function(row) {
    row > 0.1
})

wMatrix[keyFeatures[[1]],]
wMatrix[keyFeatures[[2]],]
wMatrix[keyFeatures[[3]],]

basisSamples <- apply(coef(nmfOut), 2, which.max)
basisWeights <- data.frame(basisVector = basisSamples, sample = names(basisSamples)) 
basisWeights$basisClass <- metadata$Organism[match(basisWeights$sample, metadata$File.Name)]
    
```


